### ISSUE TITLE
Questions about training my own datasets

### ISSUE THREAD

**Issue by Leiwx52** [yassouali/CCT#7]
Hi! Thank you for your contributions to this repo!

Here, I have a question about training a model for a self-collected dataset. When I used this code, I modified the argument  `self.num_classes=21` to `self.num_classes=5` and other parts which are relative to `#num_classes`, remaining the rest unchanged. However, some of the classes' IoU turns to be 0 even in the train set. Also I turned down the ''semi'' setting and used a "supervised" mode and it turned out that there was no improvement on the previous mentioned classes' IoU. 

There are two interesting facts: firstly, I trained with DeepLab-V3+ before and each of the class has non-zero IoU. Secondly, when I trained  the CCT model with `self.num_classes=21` and other hyper-parameters remain the same as in this repo, each of the class has non-zero IoU, too. That is, if I simply changed `self.num_classes`(whatever the mode is), the performance of some classes collapsed.

Can you comment on this or give some advices on how to train with a different dataset? Thx!

**Comment by yassouali** [comment]
Hi @WingsleyLui, thank you for your interest in our work;

Hmmm, weird behavior, to be honest, did you visualize/print the targets to make sure that there is no problem in the pre-processing, and the classes are present in the ground truth, and from 0 - 5. In general, make sure you have the expected results along the whole pipeline. In the beginning, only use supervised mode, if the mIoU of all classes > 0 then go to the semi model.

You can also try a adjusting the LR & more training epochs. You can also try a focal loss instead of a normal cross entropy.

**Comment by WingsleyLui** [comment]
@yassouali Thank you for your reply! 

Actually I visualized the ground-truth and the prediction using Tensorboard and as indicated by the printed info, there was no such prediction in the class of zero IoU, although there exists some annotations of that class(which means the correct pre-processing). This was showed by tensorboard in visualization of supervised images, semi-supervised images and  validation images. It also happened when I used supervised mode only. However, when the number of classes was set to 21, all the classes have non-zero IoU and the visualization of predictions looks closer to the ground-truths.

Thank you for your suggestions and I'll try different loss function. Since a smaller learning rate didn't work and this collapse happened in the very beginning of the training process, it might be not helpful to increase the number of training epochs. Apart from this, do you have other thoughts like tuning other hyper-parameters?

**Comment by WingsleyLui** [comment]
> Hi @WingsleyLui, thank you for your interest in our work;
> 
> Hmmm, weird behavior, to be honest, did you visualize/print the targets to make sure that there is no problem in the pre-processing, and the classes are present in the ground truth, and from 0 - 5. In general, make sure you have the expected results along the whole pipeline. In the beginning, only use supervised mode, if the mIoU of all classes > 0 then go to the semi model.
> 
> You can also try a adjusting the LR & more training epochs. You can also try a focal loss instead of a normal cross entropy.

**Comment by Leiwx52** [comment]
> Hi @WingsleyLui, thank you for your interest in our work;
> 
> Hmmm, weird behavior, to be honest, did you visualize/print the targets to make sure that there is no problem in the pre-processing, and the classes are present in the ground truth, and from 0 - 5. In general, make sure you have the expected results along the whole pipeline. In the beginning, only use supervised mode, if the mIoU of all classes > 0 then go to the semi model.
> 
> You can also try a adjusting the LR & more training epochs. You can also try a focal loss instead of a normal cross entropy.
@yassouali  Thank you for your reply!

Actually I visualized the ground-truth and the prediction using Tensorboard and as indicated by the printed info, there was no such prediction in the class of zero IoU, although there exists some annotations of that class(which means the correct pre-processing). This was showed by tensorboard in visualization of supervised images, semi-supervised images and validation images. It also happened when I used supervised mode only. However, when the number of classes was set to 21, all the classes have non-zero IoU and the visualization of predictions looks closer to the ground-truths.

Thank you for your suggestions and I'll try different loss function. Since a smaller learning rate didn't work and this collapse happened in the very beginning of the training process, it might be not helpful to increase the number of training epochs. Apart from this, do you have other thoughts like tuning other hyper-parameters?

**Comment by yassouali** [comment]
@WingsleyLui

If you still don't have good results, I'd suggest trying to use deeplabv3+ as the base model instead of pspnet, given that deeplabv3+ is better, just make sure to follow the same procedure, by using multiple decoders and a shared encoder, the choice of the decoders / encoder parts is up to you as a design choice.

**Comment by SuzannaLin** [comment]
I have the same issue using my own data set. I ran the framework fully supervised and still have 0.0 IoU for my classes.
![image](https://user-images.githubusercontent.com/69157347/95312316-0f999100-088f-11eb-8535-e1735d4f670d.png)
I only adjusted self.num_classes = 4 in voc.py
Is there more I should adjust?

However, when I leave self.num_classes = 21 and train on my own data set with 4 classes. The results seem good.
![image](https://user-images.githubusercontent.com/69157347/95312533-56878680-088f-11eb-84f9-e85e0dba8497.png)
Of course, the mean IoU looks low because of the 0.0 classes, so I would need to calculate mIoU myself.
